File: ios-native/ClubFuoco/Core/DesignSystem/Theme.swift
import SwiftUI

/// Design tokens lifted from the web app's inline palette
/// (BottomNav.tsx / OAuthButtons.tsx / globals.css).
enum Theme {
    // ── Colors ────────────────────────────────────────────────────────────────
    // `ink` and `cream` are a contrast *pair*: ink is the primary text color and
    // also the fill of the primary buttons/pills, whose labels are always
    // `cream`. Because both invert together, those filled controls flip to a
    // light-on-dark button in Dark Mode with no call-site change.
    static let ink = Color.adaptive(light: 0x221E1A, dark: 0xEDE6D8)       // primary text / contrast fills
    static let stone = Color.adaptive(light: 0x6E6356, dark: 0xB4AA9A)     // secondary text
    static let sand = Color.adaptive(light: 0xB0A898, dark: 0x7A7264)      // inactive / tertiary
    static let fadedSand = Color.adaptive(light: 0x9F9486, dark: 0x8A8172) // captions, overlines
    static let cream = Color.adaptive(light: 0xF8F5EE, dark: 0x0E0C0A)     // app background / labels on ink
    // Deepened in Dark. The light value is bright enough to glare against a
    // near-black background (7.3:1); the dark one lands at ~4.8:1, so it still
    // clears AA for text while sitting back in the page.
    static let gold = Color.adaptive(light: 0xC09950, dark: 0x9A7A3E)      // accent (active pill, highlights)
    static let wine = Color.adaptive(light: 0x8C2A2A, dark: 0xC85450)      // badges / destructive accents

    static let hairline = Color.adaptive(light: 0x221E1A, lightAlpha: 0.10,
                                         dark: 0xF4ECDD, darkAlpha: 0.12)

    // ── Surfaces ──────────────────────────────────────────────────────────────
    // Cards and sheets that sit *on top of* the app background. In light these
    // are the plain white the app has always used; in dark they are a step
    // lighter than the background so elevation still reads.
    static let surface = Color.adaptive(light: 0xFFFFFF, dark: 0x1A1613)
    static let surfaceRaised = Color.adaptive(light: 0xFFFFFF, dark: 0x232019)

    /// Behind async images while they load — a warm block that matches the
    /// surrounding page rather than flashing light on a dark screen.
    static let imagePlaceholder = Color.adaptive(light: 0xEFE9DD, dark: 0x231F1A)

    /// "Open now", confirmed bookings, credited fiamme. Lightened in Dark so it
    /// still separates from the background.
    static let success = Color.adaptive(light: 0x2D7A46, dark: 0x4CA96B)

    /// QR cards stay white with dark modules in BOTH modes — door scanners
    /// need the quiet zone, so these must never follow the appearance.
    static let qrSurface = Color.white
    static let onQRSurface = Color(hex: 0x221E1A)

    // Night/cinema palette (splash + dark hero surfaces, see Splash.tsx).
    // Deliberately *not* adaptive — these surfaces are dark in both modes.
    static let night = Color(hex: 0x0A0807)         // rgb(10,8,7)
    static let parchment = Color(hex: 0xF4ECDD)     // rgb(244,236,221)
    // Deepened alongside gold (4.3:1 -> ~3.2:1). Ember is an accent and a CTA
    // fill rather than body copy, and its fills carry cream/parchment labels,
    // which the deeper burnt tone actually helps.
    static let ember = Color.adaptive(light: 0xC2562D, dark: 0xA34724) // primary CTA
    static let emberCream = Color(hex: 0xFFF6E5)    // rgb(255,246,229)
    static let flame = Color(hex: 0xE8B65B)         // rgb(232,182,91) glow/badges
    static let darkRed = Color(hex: 0x6B1F1F)       // rgb(107,31,31)

    // ── Radii ─────────────────────────────────────────────────────────────────
    static let radiusField: CGFloat = 14
    static let radiusCard: CGFloat = 18
    static let radiusPill: CGFloat = 24
}

// ── Typography ────────────────────────────────────────────────────────────────
// The brand fonts, matching the web app exactly: Instrument Serif (display)
// + Geist Sans (body) + Geist Mono (kickers/labels), bundled under
// ClubFuoco/Fonts and registered via UIAppFonts in project.yml.
//
// Geist's Medium / SemiBold faces ship as their own font families, so weights
// are referenced by exact PostScript name. Instrument Serif's Regular + Italic
// share one family, so `.italic()` (here and at call sites) resolves to the
// real italic face. `fixedSize:` keeps the same non-scaling sizing the layout
// was built against (matching the previous `.system(size:)`).
extension Font {
    static func cfSerif(_ size: CGFloat, italic: Bool = false) -> Font {
        let base = Font.custom("Instrument Serif", fixedSize: size)
        return italic ? base.italic() : base
    }
    static func cfMono(_ size: CGFloat, weight: Font.Weight = .regular) -> Font {
        Font.custom(weight == .regular ? "GeistMono-Regular" : "GeistMono-Medium", fixedSize: size)
    }
    static func cfSans(_ size: CGFloat, weight: Font.Weight = .regular) -> Font {
        let name: String
        switch weight {
        case .medium:                 name = "Geist-Medium"
        case .semibold:               name = "Geist-SemiBold"
        case .bold, .heavy, .black:   name = "Geist-Bold"
        default:                      name = "Geist-Regular"
        }
        return Font.custom(name, fixedSize: size)
    }
}

/// Small uppercase monospaced overline ("kicker") used across the brand.
struct Kicker: View {
    let text: String
    var color: Color = Theme.wine
    var size: CGFloat = 9.5

    init(_ text: String, color: Color = Theme.wine, size: CGFloat = 9.5) {
        self.text = text
        self.color = color
        self.size = size
    }

    var body: some View {
        Text(text.uppercased())
            .font(.cfMono(size))
            .kerning(2.0)
            .foregroundStyle(color)
    }
}

extension Color {
    init(hex: UInt32) {
        self.init(
            red: Double((hex >> 16) & 0xFF) / 255,
            green: Double((hex >> 8) & 0xFF) / 255,
            blue: Double(hex & 0xFF) / 255
        )
    }

    /// A color that resolves per trait collection, so the ~850 `Theme.*` call
    /// sites need no change when Dark Mode flips. Built on a dynamic `UIColor`
    /// because SwiftUI has no first-class light/dark literal outside an asset
    /// catalog.
    static func adaptive(light: UInt32, lightAlpha: Double = 1,
                         dark: UInt32, darkAlpha: Double = 1) -> Color {
        Color(uiColor: UIColor { traits in
            traits.userInterfaceStyle == .dark
                ? UIColor(hex: dark, alpha: darkAlpha)
                : UIColor(hex: light, alpha: lightAlpha)
        })
    }
}

private extension UIColor {
    convenience init(hex: UInt32, alpha: Double) {
        self.init(
            red: CGFloat((hex >> 16) & 0xFF) / 255,
            green: CGFloat((hex >> 8) & 0xFF) / 255,
            blue: CGFloat(hex & 0xFF) / 255,
            alpha: CGFloat(alpha)
        )
    }
}

/// Haptics on key actions — part of the "feels native" goal.
enum Haptics {
    @MainActor static func tap() {
        UIImpactFeedbackGenerator(style: .light).impactOccurred()
    }
    @MainActor static func success() {
        UINotificationFeedbackGenerator().notificationOccurred(.success)
    }
    @MainActor static func error() {
        UINotificationFeedbackGenerator().notificationOccurred(.error)
    }
}

File: android-native/app/src/main/java/com/clubfuoco/app/core/designsystem/Theme.kt
package com.clubfuoco.app.core.designsystem

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.ReadOnlyComposable
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.clubfuoco.app.R

/**
 * Design tokens — a direct port of ios-native's `Theme.swift`, which in turn
 * lifted them from the web app's palette. The two files must stay in step;
 * change one, change the other.
 *
 * The iOS version leans on `Color.adaptive(light:dark:)`, a dynamic UIColor
 * that resolves per trait collection so ~850 call sites need no branching.
 * Compose has no equivalent, so the same trick is done with a palette object
 * chosen once at the top of the tree and read through a CompositionLocal:
 * `Theme.ink` still means "the right ink for the current appearance" at every
 * call site, and nothing downstream has to know which mode is active.
 */
data class FuocoPalette(
    // `ink` and `cream` are a contrast PAIR: ink is the primary text colour and
    // also the fill of primary buttons/pills, whose labels are always cream.
    // Because both invert together, those filled controls flip to light-on-dark
    // in dark mode with no call-site change.
    val ink: Color,
    val stone: Color,
    val sand: Color,
    val fadedSand: Color,
    val cream: Color,
    val gold: Color,
    val wine: Color,
    /**
     * Wine's *decorative* half — venue names, "6 venues →" links, section
     * icons, stat numbers. On the cream page it is the same red; against
     * near-black a page full of red reads as noise, so it goes off-white and
     * lets the type carry the hierarchy.
     *
     * Keep [wine] itself for red that MEANS something: errors, destructive
     * actions, badges. Those stay red in both modes.
     */
    val accent: Color,
    val hairline: Color,
    val surface: Color,
    val surfaceRaised: Color,
    val imagePlaceholder: Color,
    val success: Color,
    val ember: Color,
    val flame: Color,
)

private val LightPalette = FuocoPalette(
    ink = Color(0xFF221E1A),
    stone = Color(0xFF6E6356),
    sand = Color(0xFFB0A898),
    fadedSand = Color(0xFF9F9486),
    cream = Color(0xFFF8F5EE),
    gold = Color(0xFFC09950),
    wine = Color(0xFF8C2A2A),
    accent = Color(0xFF8C2A2A),
    hairline = Color(0x1A221E1A),
    surface = Color(0xFFFFFFFF),
    surfaceRaised = Color(0xFFFFFFFF),
    imagePlaceholder = Color(0xFFEFE9DD),
    success = Color(0xFF2D7A46),
    ember = Color(0xFFC2562D),
    flame = Color(0xFFE8B65B),
)

// Saturation, not hue, is what makes the golds read as gold. A near-pure hue
// looks like paint at any lightness — yellow up around 44°, orange down at 35° —
// and chasing it by moving hue just trades one for the other. The brand's light
// gold sits at S .58, and that muting is the whole trick, so the dark variants
// hold hue ~41° and match the saturation, taking the deepening in value alone.
private val DarkPalette = FuocoPalette(
    ink = Color(0xFFEDE6D8),
    stone = Color(0xFFB4AA9A),
    sand = Color(0xFF7A7264),
    fadedSand = Color(0xFF8A8172),
    cream = Color(0xFF0E0C0A),
    gold = Color(0xFFAD8A45),
    // The first dark wine lifted brightness without saturation and came out
    // coral. #B33F38 holds the red hue at S .69 instead of .60 and drops
    // V .78 → .70, so it reads as wine again.
    wine = Color(0xFFB33F38),
    accent = Color(0xFFE8E0D2),
    hairline = Color(0x1FF4ECDD),
    surface = Color(0xFF1A1613),
    surfaceRaised = Color(0xFF232019),
    imagePlaceholder = Color(0xFF231F1A),
    success = Color(0xFF4CA96B),
    // Saturation up rather than value down, so it lands bolder instead of
    // muddy — an orange that still has bite against the near-black page.
    ember = Color(0xFFB8431A),
    flame = Color(0xFFC7A150),
)

/**
 * Fixed colours — deliberately NOT adaptive, because the surface they sit on is
 * dark (or must stay light) in both appearances.
 */
object FuocoFixed {
    /** Splash + cinema hero surfaces. */
    val night = Color(0xFF0A0807)
    val parchment = Color(0xFFF4ECDD)
    val emberCream = Color(0xFFFFF6E5)
    val darkRed = Color(0xFF6B1F1F)

    /**
     * QR cards stay white with dark modules in BOTH modes — door scanners need
     * the quiet zone, so these must never follow the appearance.
     */
    val qrSurface = Color.White
    val onQrSurface = Color(0xFF221E1A)

    /**
     * The rating star over photo cards. Sits on a scrim over a photo, which is
     * dark either way, so it is fixed at H40 / S.63 — gold rather than the
     * highlighter-yellow it used to be.
     */
    val starGold = Color(0xFFE0B052)
}

object FuocoRadius {
    val field = 14.dp
    val card = 18.dp
    val pill = 24.dp
}

// ── Typography ───────────────────────────────────────────────────────────────
// The brand fonts, matching the web and iOS apps exactly: Instrument Serif
// (display) + Geist Sans (body) + Geist Mono (kickers/labels), copied from
// ios-native/ClubFuoco/Fonts into res/font.
//
// Geist ships Medium/SemiBold/Bold as separate files, which Compose models as
// weights inside one family — cleaner than iOS, where they are separate
// families referenced by PostScript name.

val InstrumentSerif = FontFamily(
    Font(R.font.instrumentserif_regular, FontWeight.Normal),
    Font(R.font.instrumentserif_italic, FontWeight.Normal, FontStyle.Italic),
)

val Geist = FontFamily(
    Font(R.font.geist_regular, FontWeight.Normal),
    Font(R.font.geist_medium, FontWeight.Medium),
    Font(R.font.geist_semibold, FontWeight.SemiBold),
    Font(R.font.geist_bold, FontWeight.Bold),
)

val GeistMono = FontFamily(
    Font(R.font.geistmono_regular, FontWeight.Normal),
    Font(R.font.geistmono_medium, FontWeight.Medium),
)

private val LocalFuocoPalette = staticCompositionLocalOf { LightPalette }

/**
 * The app's palette at the current appearance. Named `Theme` so ported Swift
 * reads almost unchanged: `Theme.ink` here is `Theme.ink` there.
 */
object Theme {
    val colors: FuocoPalette
        @Composable @ReadOnlyComposable get() = LocalFuocoPalette.current

    val ink: Color @Composable @ReadOnlyComposable get() = colors.ink
    val stone: Color @Composable @ReadOnlyComposable get() = colors.stone
    val sand: Color @Composable @ReadOnlyComposable get() = colors.sand
    val fadedSand: Color @Composable @ReadOnlyComposable get() = colors.fadedSand
    val cream: Color @Composable @ReadOnlyComposable get() = colors.cream
    val gold: Color @Composable @ReadOnlyComposable get() = colors.gold
    val wine: Color @Composable @ReadOnlyComposable get() = colors.wine
    val accent: Color @Composable @ReadOnlyComposable get() = colors.accent
    val hairline: Color @Composable @ReadOnlyComposable get() = colors.hairline
    val surface: Color @Composable @ReadOnlyComposable get() = colors.surface
    val surfaceRaised: Color @Composable @ReadOnlyComposable get() = colors.surfaceRaised
    val imagePlaceholder: Color @Composable @ReadOnlyComposable get() = colors.imagePlaceholder
    val success: Color @Composable @ReadOnlyComposable get() = colors.success
    val ember: Color @Composable @ReadOnlyComposable get() = colors.ember
    val flame: Color @Composable @ReadOnlyComposable get() = colors.flame
}

/**
 * Root theme wrapper. [dark] is passed explicitly (rather than always reading
 * the system) because appearance is user-controlled in Settings — Light / Dark /
 * System — exactly as `ThemeStore` drives `preferredColorScheme` on iOS.
 */
@Composable
fun ClubFuocoTheme(
    dark: Boolean = isSystemInDarkTheme(),
    content: @Composable () -> Unit,
) {
    CompositionLocalProvider(
        LocalFuocoPalette provides if (dark) DarkPalette else LightPalette,
        content = content,
    )
}
